### PR TITLE
NotRequired needs typing_extensions >=4.0.0

### PR DESCRIPTION
**Pull request by kernc**
- [ ] issues: fixes https://github.com/bokeh/bokeh/issues/12537
- [ ] tests added / passed

### CODE REVIEW

File: src/bokeh/core/serialization.py [hunk 1]
```diff
@@ -65,6 +64,8 @@
 from .types import ID
 
 if TYPE_CHECKING:
+    from typing_extensions import NotRequired, TypeAlias
```
Comment (bryevdv) [inline src/bokeh/core/serialization.py:67]: would it make sense to do this thoughout the codebase?

Comment (bryevdv) [reply src/bokeh/core/serialization.py:67]: @mattpap If we did that we could potentially drop `typing_extensions` as a *runtime* requirement entirely, no?

Comment (kernc) [reply src/bokeh/core/serialization.py]: If the checks pass, I'll let you decide.

While I have your attention, would you have any idea on the issue with https://github.com/kernc/backtesting.py/actions/runs/3301867863/jobs/5447910367#step:9:3? :sweat_smile:

Comment (bryevdv) [reply src/bokeh/core/serialization.py]: @kernc  looks like some migrations to handle, let's take those discussions up on the discourse https://discourse.bokeh.org/

**Comment by bryevdv** [comment]
@kernc BTW I note that you have submitted several PRs now. Please let me know of you would like to be added to `@bokeh/dev` Primarily that means that you can push PR branches directly to the main repo, which can make collaboration a little more seamless, and you could be notified when a question or comment is addressed to `@bokeh/dev`

**Comment by kernc** [comment]
Haaaa ... 😅 I sure like and have interest in the software, but I'm already having obvious issues with keeping up with obligations for scikit-optimize I promised to maintain, not to mention my own packages.

At least for the time being, I'd rather just keep them coming. But thanks!

**Comment by kernc** [comment]
What's best to do about this [`isort` failing](https://github.com/bokeh/bokeh/actions/runs/3372830985/jobs/5596863648#step:5:58)?

**Comment by bryevdv** [comment]
@kernc you can try running `isort -rc` in the respecitive top-level subdirs (not at the very top level) or I can try to fix up and push changes to your branch tonight

**Comment by kernc** [comment]
What I don't like is isort stripping the relevant comment line. If we're to blindly defer decision to an uninformed tool, I'd rather you do it. :grin:

**Comment by bryevdv** [comment]
> What I don't like is isort stripping the relevant comment line. 


We actually consider anything under `"if TYPE_CHECKING` to be "code" rather than imports, so it's not categorized with all the other normal imports. The alternative would be to possibly necessitate multiple `"if TYPE_CHECKING` blocks under each section (e.g. if some type imports were stdlib, others external, and others pacakge local) That would get tedious and ugly fast :)

**Comment by bryevdv** [comment]
FYI integration test is flaky, failure not related to this PR

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
Thanks for the PR @kernc I know it grew a little past initial estimates. I will merge this now and follow-up with a PR to remove the `typing_extensions` dependency altogether, with a test to maintain that `typing_extensions` is not imported anywhere.

**Pull request merged**